# Elastic Net Regression

* The aim is to find the coefficients of independent variables that minimize the Root Mean Square Error(RMSE) by applying a penalty to these coefficients.



* Elastis Net Regression combines the lasso and Ridge regression approach.



* Elastis Net Regression Is the most advanced form of regression models

## 1-)Data Preprocessing

In [1]:
import pandas as pd
hit = pd.read_csv("Hitters.csv")
df = hit.copy()
df = df.dropna()
dms = pd.get_dummies(df[['League', 'Division', 'NewLeague']])
y = df["Salary"]
X_ = df.drop(['Salary', 'League', 'Division', 'NewLeague'], axis=1).astype('float64')
X = pd.concat([X_, dms[['League_N', 'Division_W', 'NewLeague_N']]], axis=1)


In [2]:
from sklearn.model_selection import train_test_split

In [3]:
X_train,X_test, y_train, y_test = train_test_split(X, 
                                                    y, 
                                                    test_size=0.25, 
                                                    random_state=42)

print("X_train", X_train.shape)

print("y_train",y_train.shape)

print("X_test",X_test.shape)

print("y_test",y_test.shape)

training = df.copy()

print("training", training.shape)

X_train (197, 19)
y_train (197,)
X_test (66, 19)
y_test (66,)
training (263, 20)


## 2-)Model

In [4]:
from sklearn.linear_model import ElasticNet

In [5]:
enet_model = ElasticNet().fit(X_train, y_train)

C:\Users\Dell\Anaconda3\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:531: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 8487984.799367279, tolerance: 3898.686956380658
  positive)


In [6]:
enet_model.coef_ # the coeeficeints of independent variables BEFORE MODEL TUNING

array([ -1.86256172,   8.70489065,   5.10426375,  -2.89875799,
        -1.28642985,   5.24343682,   6.04480276,  -0.14701495,
        -0.21566628,  -0.7897201 ,   1.80813117,   0.80914508,
        -0.61262382,   0.26816203,   0.27172387,  -0.36530729,
        19.2186222 , -31.16586592,   8.98369938])

In [7]:
enet_model.intercept_

-6.465955602113127

## 3-) Prediction

In [12]:
enet_model

ElasticNet()

In [13]:
y_pred = enet_model.predict(X_test)
y_pred[0:5]

array([ 577.79111731,  617.33202224, 1031.39113156,  364.95861575,
        489.51894393])

In [18]:
import numpy as np
from sklearn.metrics import mean_squared_error,r2_score

In [19]:
test_error_before=np.sqrt(mean_squared_error(y_test, y_pred))
test_error_before #test error before model tuning

357.1676548181244

In [20]:
r2_score(y_test, y_pred)

0.41070222469326945

## 4-)Model Tuning

In [21]:
from sklearn.linear_model import ElasticNetCV

* We will try to find the optimum alpha value using the cros validation method.




* In other words, instead of choosing the alpha value randomly ,which is called a hyperparameter we will find it with the cross validation method.



* Finally, we will create the final model by using the most optimum alpha value.


In [23]:
enet_model.alpha # the alpha value before model tuning

1.0

### 4.1-) Croos Validation Method

In [24]:
enet_cv_model = ElasticNetCV(cv = 10, random_state = 0).fit(X_train, y_train)

In [26]:
enet_cv_model.alpha_ #the most optimum alpha value 

5230.7647364798695

### 4-2) Fitting final model by using optimum alpha value

In [27]:
enet_tuned = ElasticNet(alpha = enet_cv_model.alpha_).fit(X_train,y_train)

In [29]:
y_pred1 = enet_tuned.predict(X_test)
y_pred1[0:5]

array([667.85450144, 550.2991971 , 853.94283928, 353.08196143,
       606.28730961])

In [30]:
test_score_after=np.sqrt(mean_squared_error(y_test, y_pred1))
test_score_after #test error after model tuning

394.15280563218795

### 5-) Comparison of test errors of tuned models

* Elastic Net Regression Test Error ==== >>>> 394.15280563218795


* Ridge Regression Test Error ==== >>>> 386.68264297564144


* Lasso Regression Test Error ==== >>>> 356.52263769583686

* Since the lowest test score belongs to Lasso Regression among these linear models, the most successful model is the Lasso Regression model.



* Therefore, we can form the equation of best model by using the coefficients of independent variables belonging to Lasso Regression Model .

